controllers/nnf_workflow_controller.go

```diff
@@ -206,132 +193,182 @@ func (r *NnfWorkflowReconciler) updateDriversStatusForStatusState(workflow *dwsv
 		if workflow.Status.State != driverStatus.WatchState {
 			continue
 		}
-		if workflow.Status.Ready {
-			continue
-		}
 		if driverStatus.Completed {
 			continue
 		}
 
-		ts := metav1.NowMicro()
-		driverStatus.CompleteTime = &ts
-		driverStatus.Completed = true
-		updateWorkflow = true
+		driverList = append(driverList, driverStatus)
 	}
 
-	return updateWorkflow
-}
-
-func (r *NnfWorkflowReconciler) completeDriverState(ctx context.Context, workflow *dwsv1alpha1.Workflow, driverID string, log logr.Logger) (ctrl.Result, error) {
+	startFunctions := map[string]func(*NnfWorkflowReconciler, context.Context, *dwsv1alpha1.Workflow, int) (*ctrl.Result, error){
+		dwsv1alpha1.StateProposal.String(): (*NnfWorkflowReconciler).startProposalState,
+		dwsv1alpha1.StateSetup.String():    (*NnfWorkflowReconciler).startSetupState,
+		dwsv1alpha1.StateDataIn.String():   (*NnfWorkflowReconciler).startDataInOutState,
+		dwsv1alpha1.StatePreRun.String():   (*NnfWorkflowReconciler).startPreRunState,
+		dwsv1alpha1.StatePostRun.String():  (*NnfWorkflowReconciler).startPostRunState,
+		dwsv1alpha1.StateDataOut.String():  (*NnfWorkflowReconciler).startDataInOutState,
+		dwsv1alpha1.StateTeardown.String(): (*NnfWorkflowReconciler).startTeardownState,
+	}
 
-	driverUpdate := r.updateDriversStatusForStatusState(workflow, driverID, log)
-	if driverUpdate {
-		if err := r.Update(ctx, workflow); err != nil {
-			if apierrors.IsConflict(err) {
-				return ctrl.Result{Requeue: true}, nil
+	// Call the correct "start" function based on workflow state for each directive that has registered for
+	// it. The "start" function does the initial work of setting up and creating the appropriate child resources.
+	for _, driverStatus := range driverList {
+		log.Info("Start", "State", workflow.Status.State, "index", driverStatus.DWDIndex, "directive", workflow.Spec.DWDirectives[driverStatus.DWDIndex])
+		result, err := startFunctions[workflow.Status.State](r, ctx, workflow, driverStatus.DWDIndex)
+		if err != nil {
+			e, ok := err.(*nnfv1alpha1.WorkflowError)
+			if ok {
+				driverStatus.Message = e.GetMessage()
+				if e.GetRecoverable() {
+					driverStatus.Reason = "running"
+				} else {
+					driverStatus.Reason = "error"
+				}
+			} else {
+				driverStatus.Reason = "error"
+				driverStatus.Message = "Internal error: " + err.Error()
 			}
 
-			log.Error(err, "Failed to update Workflow state")
+			log.Info("Start error", "State", workflow.Status.State, "index", driverStatus.DWDIndex)
```

Comment:
Log the error here.   I'm hitting this line in my test runs, and I'm having trouble figuring out where the error message is in the test output.